pkg/distributor/distributor.go

```diff
@@ -207,19 +248,23 @@ func (d *Distributor) Push(ctx context.Context, req *connect.Request[pushv1.Push
 		for _, raw := range series.Samples {
 			usagestats.NewCounter(fmt.Sprintf("distributor_profile_type_%s_received", profName)).Inc(1)
 			d.profileReceivedStats.Inc(1)
-			d.bytesReceivedTotalStats.Inc(int64(len(raw.RawProfile)))
-			d.bytesReceivedStats.Record(float64(len(raw.RawProfile)))
+			if haveRawPprof {
+				d.metrics.receivedCompressedBytes.WithLabelValues(profName, tenantID).Observe(float64(len(raw.RawProfile)))
+			}
 			totalProfiles++
-			d.metrics.receivedCompressedBytes.WithLabelValues(profName, tenantID).Observe(float64(len(raw.RawProfile)))
-			p, err := pprof.RawFromBytes(raw.RawProfile)
-			if err != nil {
-				return nil, connect.NewError(connect.CodeInvalidArgument, err)
+			p := raw.Profile
+			var decompressedSize int
+			if haveRawPprof {
+				decompressedSize = p.SizeBytes()
+			} else {
+				decompressedSize = p.SizeVT()
 			}
-			d.metrics.receivedDecompressedBytes.WithLabelValues(profName, tenantID).Observe(float64(p.SizeBytes()))
+			d.metrics.receivedDecompressedBytes.WithLabelValues(profName, tenantID).Observe(float64(decompressedSize))
 			d.metrics.receivedSamples.WithLabelValues(profName, tenantID).Observe(float64(len(p.Sample)))
-			totalPushUncompressedBytes += int64(p.SizeBytes())
+			totalPushUncompressedBytes += int64(decompressedSize)
 
-			if err := validation.ValidateProfile(d.limits, tenantID, p.Profile, p.SizeBytes(), phlaremodel.Labels(series.Labels)); err != nil {
+			if err := validation.ValidateProfile(d.limits, tenantID, p.Profile, decompressedSize, series.Labels); err != nil {
+				//todo this actually discards more if multiple Samples in a Series request
```

Comment:
Correct, right now the Grafana Agents never send more than multiple sample per series.

I also think this needs better care in the future.